kotlinx-coroutines-core/common/src/selects/Select.kt

```diff
@@ -342,69 +358,245 @@ internal class SelectBuilderImpl<in R>(
+       7. SelectBuilderImpl.trySelectOther creates PairSelectOp(7) and installs it as SelectBuilderImpl(1).state(c)
+          to secure the state of the first builder and commit ability to make it selected for this operation.
+       8. NOW THE RENDEZVOUS IS FULLY PREPARED via descriptors installed at
+          - SelectBuilderImpl(2).state(a)
+          - Send/ReceiveSelect(3).next(b)
+          - SelectBuilderImpl(1).state(c)
+          Any concurrent operation that is trying to access any of the select instances or the queue is going to help.
+          Any helper that helps AtomicSelectOp(4) calls TryPoll/OfferDesc(5).prepare which tries to determine
+          "affectedNode" but is bound to discover the same Send/ReceiveSelect(3) node that cannot become
+          non-first node until this operation completes (there are no insertions to the head of the queue!)
+          We have not yet decided to complete this operation, but we cannot ever decide to complete this operation
+          on any other node but Send/ReceiveSelect(3), so it is now safe to perform the next step.
+       9. PairSelectOp(7).perform calls PrepareOp(6).finishPrepare which copies PrepareOp(6).affected and PrepareOp(6).next
+          to the corresponding TryPoll/OfferDesc(5) fields.
+       10. PairSelectOp(7).perform calls AtomicSelect(4).decide to reach consensus on successful completion of this
+          operation. This consensus is important in light of dead-lock resolution algorithm, because a stale helper
+          could have stumbled upon a higher-numbered atomic operation and had decided to abort this atomic operation,
+          reaching decision on RETRY_ATOMIC status of it. We cannot proceed with completion in this case and must abort,
+          all objects including AtomicSelectOp(4) will be dropped, reverting all the three updated pointers to
+          their original values and atomic operation will retry from scratch.
+       11. NOW WITH SUCCESSFUL UPDATE OF AtomicSelectOp(4).consensus to null THE RENDEZVOUS IS COMMITTED. The rest
+           of the code proceeds to update:
+           - SelectBuilderImpl(1).state to TryPoll/OfferDesc(5) so that late helpers would know that we have
+             already successfully completed rendezvous.
+           - Send/ReceiveSelect(3).next to Removed(next) so that this node becomes marked as removed.
+           - SelectBuilderImpl(2).state to null to mark this select instance as selected.
+
+       Note, that very late helper may try to perform this AtomicSelectOp(4) when it is already completed.
+       It can proceed as far as finding affected node, creating PrepareOp, installing this new PrepareOp into the
+       node's next pointer, but PrepareOp.perform checks that AtomicSelectOp(4) is already decided and undoes all
+       the preparations.
+     */
+
+    // it is just like plain trySelect, but support idempotent start
```

Comment:
Explanation and state diagram is outstanding, though I don't think I am able to find any bugs here if they exist anyway, the whole select machinery becomes more and more complex to keep it in the head during reasoning.

Stress-tests as a palliative and giving up on the lock-freedom in new channels looks like a way to go for now.

Reply:
Totally agree. There's no way to reason about correctness of this implementation. It is too complex.